skorch/classifier.py

```diff
@@ -188,3 +189,160 @@ def predict(self, X):
+        super().check_data(X, y)
+        if get_dim(y) != 1:
+            raise ValueError("The target data should be 1-dimensional.")
+
+    # pylint: disable=signature-differs
+    def fit(self, X, y, **fit_params):
+        """See ``NeuralNet.fit``.
+
+        In contrast to ``NeuralNet.fit``, ``y`` is non-optional to
+        avoid mistakenly forgetting about ``y``. However, ``y`` can be
+        set to ``None`` in case it is derived dynamically from
+        ``X``.
+
+        """
+        # pylint: disable=useless-super-delegation
+        # this is actually a pylint bug:
+        # https://github.com/PyCQA/pylint/issues/1085
+        return super().fit(X, y, **fit_params)
+
+    def predict(self, X, threshold=0.5):
+        """Where applicable, return class labels for samples in X.
+
+        If the module's forward method returns multiple outputs as a
+        tuple, it is assumed that the first output contains the
+        relevant information and the other values are ignored. If all
+        values are relevant, consider using
+        :func:`~skorch.NeuralNet.forward` instead.
+
+        Parameters
+        ----------
+        X : input data, compatible with skorch.dataset.Dataset
+          By default, you should be able to pass:
+
+            * numpy arrays
+            * torch tensors
+            * pandas DataFrame or Series
+            * a dictionary of the former three
+            * a list/tuple of the former three
+            * a Dataset
+
+          If this doesn't work with your data, you have to pass a
+          ``Dataset`` that can deal with the data.
+
+        Returns
+        -------
+        y_pred : numpy ndarray
+
+        """
+        return (self.predict_proba(X) > threshold).astype('uint8')
+
+    # pylint: disable=missing-docstring
+    def predict_proba(self, X):
+        """Where applicable, return probability estimates for
+        samples.
+
+        If the module's forward method returns multiple outputs as a
+        tuple, it is assumed that the first output contains the
+        relevant information and the other values are ignored. If all
+        values are relevant, consider using
+        :func:`~skorch.NeuralNet.forward` instead.
+
+        Parameters
+        ----------
+        X : input data, compatible with skorch.dataset.Dataset
+          By default, you should be able to pass:
+
+            * numpy arrays
+            * torch tensors
+            * pandas DataFrame or Series
+            * a dictionary of the former three
+            * a list/tuple of the former three
+            * a Dataset
+
+          If this doesn't work with your data, you have to pass a
+          ``Dataset`` that can deal with the data.
+
+        Returns
+        -------
+        y_proba : numpy ndarray
+
+        """
+        y_probas = []
+        for yp in self.forward_iter(X, training=False):
+            yp = yp[0] if isinstance(yp, tuple) else yp
+            yp = torch.nn.functional.sigmoid(yp)
```

Comment:
As with the multi-class classifier the module should return what the user wants to see in `predict_proba`. Consequently, I don't think that the call to sigmoid is justified here. Maybe it would be better to expect values in the range from 0 to 1 from the module.

Reply:
I decided to use `BCEWithLogitsLoss` as the loss function, which expects the module to output the logit of probability with a range of (-inf, inf). I choose this loss function because it is more numerically stable than `BCELoss`.

Since logit of probability is the module output, sigmoid will convert it to a probability.

Reply:
I see why you made that decision but wonder if we need a check for the type of loss function used like in [NeuralNetClassifier](https://github.com/dnouri/skorch/blob/12842307309765b8a0caefa00441df88ec53ce0d/skorch/classifier.py#L103).

Reply:
`NeutralNetClassifier` assumes that the module output is probabilities and `NLLLoss` expects log probabilities, thus that check and log call needs to be in place.

`NeuralNetBinaryClassifier` assumes that the module outputs logits, and `BCEWithLogitsLoss` expects logits, thus no checks need to be in place.

I see how this can confuse a user, where `NeutralNetClassifier` and `NeuralNetBinaryClassifier` expects different outputs from the module. 

It is a choice that needs to be made:

1. Have both `NeutralNet*Classifier`s expect the module to output probabilities for a cleaner API, at the cost of `BCEWithLogitsLoss` using `BCELoss`.
2. Have `NeutralNetClassifier` expect the module to output probabilities, and `NeuralNetBinaryClassifier` expect logits. This allows `NeuralNetBinaryClassifier` to use `BCEWithLogitsLoss` to be numerically stable. The cost would be a less clean API.

Reply:
Ah I see what you mean. I updated this PR with your suggestion.